csrc/scheduler/mma_utils.cpp

```diff
@@ -14,11 +15,37 @@
 #include <scheduler/utils.h>
 #include <variant>
 #include "mma_type.h"
-
 namespace nvfuser {
 
 namespace mma_utils {
 
+bool hasEnoughSharedMemoryForEpilogue(
+    const MatMulTileOptions& gemm_tile,
+    const int smem_double_buffer_stage) {
+  auto properties = at::cuda::getDeviceProperties(
+      c10::Device(c10::DeviceType::CUDA, 0).index());
+  const size_t device_smem_limit = properties->sharedMemPerBlockOptin;
+
+  // see scheduleContiguousVectorLoad
+  const int vector_word = 8;
+  auto warp_dims = gemm_tile.cta_tile / gemm_tile.warp_tile;
+  const int round_to_factor =
+      warp_dims.m * warp_dims.n * warp_dims.k * 32 * vector_word;
+  const int mk = gemm_tile.cta_tile.m * gemm_tile.cta_tile.k;
+  const int nk = gemm_tile.cta_tile.n * gemm_tile.cta_tile.k;
+  const size_t smem_a = (size_t)(ceilDiv(mk, round_to_factor) *
+                                 round_to_factor * smem_double_buffer_stage) *
+      dataTypeSize(DataType::Half);
+  const size_t smem_b = (size_t)(ceilDiv(nk, round_to_factor) *
+                                 round_to_factor * smem_double_buffer_stage) *
+      dataTypeSize(DataType::Half);
+  const size_t smem_c = (size_t)(gemm_tile.cta_tile.m * gemm_tile.cta_tile.n) *
+      dataTypeSize(DataType::Float);
+  const size_t smem_size = smem_a + smem_b + smem_c;
+
+  return smem_size <= device_smem_limit;
```

Comment:
I think you are checking if `smem_size <= device_smem_limit` and enable smem unswizzle if this condition is satisfied. I don't think this is the correct behavior. Each kernel has its designed occupancy, if the designed occupancy is `N`, then we should have `N*smem_size <= device_smem_limit`. cc: @drzejan2 Any opinion on this?

Reply:
You are right @zasdfgbnm

The first though that comes to my mind is to calculate current occupancy where only `smem_a` and `smem_b` are in use and then check if new occupancy, with `smem_a`, `smem_b` and `smem_c`, does not change the occupancy. Something along these lines:

```
const auto curr_smem_occupancy = device_smem_limit / (smem_a + smem_b);
const auto new_smem_occupancy = device_smem_limit / (smem_a + smem_b + smem_c);

return new_smem_occupancy <= curr_smem_occupancy;
```

Reply:
Good point. But without memory reuse `new_smem_occupancy < curr_smem_occupancy` is almost always true. e.g. I checked `NVFuserTest.FusionAmpereMatmulTileCheck4warp_CUDA` and the results are:
```
curr_smem_occupancy= 20 new_smem_occupancy= 13
curr_smem_occupancy= 10 new_smem_occupancy= 8
curr_smem_occupancy= 10 new_smem_occupancy= 8
curr_smem_occupancy= 10 new_smem_occupancy= 5
curr_smem_occupancy= 6 new_smem_occupancy= 4
curr_smem_occupancy= 5 new_smem_occupancy= 3
curr_smem_occupancy= 6 new_smem_occupancy= 2
curr_smem_occupancy= 4 new_smem_occupancy= 2
curr_smem_occupancy= 3 new_smem_occupancy= 1
curr_smem_occupancy= 5 new_smem_occupancy= 1
curr_smem_occupancy= 3 new_smem_occupancy= 1
curr_smem_occupancy= 2 new_smem_occupancy= 1
curr_smem_occupancy= 4 new_smem_occupancy= 1
curr_smem_occupancy= 2 new_smem_occupancy= 0
```
This function was designed to ensure we can launch the kernel with shared memory epilogue without considering the occupancy. By doing this we can increase the number of cases can be tested with smem epilogue. For performance purpose we definitly need `N*smem_size <= device_smem_limit`, in this revision N is derived from register limitation. For case `NVFuserTest.FusionAmpereMatmulTileCheck4warp_CUDA` N is 1 or 2.
```
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 1 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 1 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 1 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 2
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 1
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 2 blocks_per_sm_with_smem_epilogue= 0
blocks_per_sm_without_smem_epilogue= 1 blocks_per_sm_with_smem_epilogue= 0
```